### ISSUE TITLE
Stale file content reused internally after file autocorrected

### ISSUE THREAD

**Issue by 3flex** [detekt/detekt#9379]
## Expected Behavior
Modified text (after autocorrect) is used when expected.

## Observed Behavior
Original file content (with a rule violation that can be autocorrected) is read in certain instances when modified text (after autocorrect) is expected.

## Steps to Reproduce
Claude has identified this issue when I asked it to look for problems in KtLintEngine.

Result of issue is showing up in CI test runs for Kotlin 2.4.0 update, for example:
* https://github.com/detekt/detekt/actions/runs/27181867404/job/80300094140?pr=9218
* https://github.com/detekt/detekt/actions/runs/27181867404/job/80300094096?pr=9218

## Context
#9291 by @e5l introduced performance improvements for the ktlint wrapper, however it seems to have introduced at least one regression.

This test was proposed by Claude to demonstrate the problem. It fails on `main` but passes when reverting #9291:
```kotlin
    @Test
    fun `a later visit on a reused KtFile must not be served a stale context after modifiedText changes`() {
        // Trigger a second, unrelated rule on a different file so the engine's registry holds two
        // active rules. This guarantees remainingVisits starts at >= 2 regardless of test ordering.
        val otherFile = compileContentForTest("fun other() = Unit\n", "EngineStaleOther.kt")
        NoSemicolons(Config.empty).visitFile(otherFile, FakeLanguageVersionSettings())

        val ktFile = compileContentForTest("fun main()\n= Unit\n", "EngineStaleContext.kt")
        val rule = NoLineBreakBeforeAssignment(Config.empty)

        // First pass: the file has a violation and a single finding is expected.
        assertThat(rule.visitFile(ktFile, FakeLanguageVersionSettings())).hasSize(1)

        // Simulate a prior autocorrect having cleaned the file.
        ktFile.modifiedText = "fun main() = Unit\n"

        // Second pass: this should read the modified (now clean) file and report nothing.
        val secondPass = rule.visitFile(ktFile, FakeLanguageVersionSettings())

        assertThat(secondPass).isEmpty()
    }
```

Claude identified two defects in the KtLintEngine implementation, and proposed some fixes. I have next to no experience with concurrent code so I'm not willing to open a PR with the proposed fixes.

Root cause analysis from Claude. I can't confirm this is correct.
<details>

## Summary

The flakiness was introduced by the recent perf refactor #9291 (`95e298a`, "collapse N walks per file into one"). `KtlintEngine` is a global `object` singleton holding **mutable state that is never reset between tests** — and it's combined with IntelliJ's PSI caching, which hands back the *same* `KtFile` instance to every test that loads the same resource. Two concrete defects fall out of that.

## Defect 1 — the eviction counter can't reach zero, so stale contexts get served

`KtlintEngine.kt:108` builds the per-file context with:

```kotlin
return FileContext(fileCopy, findings, AtomicInteger(active.size))
```

`remainingVisits` is seeded with **`active.size`** — the number of rules that *participated in the shared walk*. But it's decremented (`KtlintEngine.kt:63-67`) once per rule that actually calls `visit()` on the file:

```kotlin
fun ruleDoneWithFile(root: KtFile, context: FileContext) {
    if (context.remainingVisits.decrementAndGet() <= 0) {
        contexts.remove(root)
    }
}
```

Those two counts are not the same. `active` is computed from the **global `registry`** (line 87), which accumulates every rule whose `visit()` has *ever* run in the JVM (each permanently sets `triggeredVisit = true`, `KtlintRule.kt:60-67`). A typical test only invokes `.lint()` on one or two rules, so:

- `active.size` = e.g. 12 (everything triggered by earlier tests), but
- only 1 rule calls `ruleDoneWithFile` → counter goes 12 → 11, **never hits 0**.

So the context is **never evicted**. Now combine with PSI caching: `loadFile("configTests/fixed.kt")` → `KtTestCompiler.compile` → `PsiUtilCore.getPsiFile(project, virtualFile)` against a **shared singleton project** (`KtTestCompiler.kt:22,35`) returns the *same* `KtFile` instance every time. That instance is the cache key in `contexts`. The next test calling `contextFor(sameFile)` gets the **stale FileContext from a previous test** — built under the previous test's config and text — instead of rebuilding. The `testFile.modifiedText = null` reset in `AutoCorrectLevelSpec.kt:73-74` (note the telltale comment *"otherwise it will be persisted between tests"*) does nothing, because the engine never re-reads `modifiedText` — it just returns the cached context. **Result: order-dependent, intermittent failures**, plus the PSI memory leak the KDoc explicitly claims to avoid.

## Defect 2 — the global registry is never cleared, so stale rules mutate unrelated files

Because `registry` (`KtlintEngine.kt:40`) is a permanent global with no clear/reset, rules from earlier tests (all with `triggeredVisit = true`) are included in `active` for *every later file* (line 87). They run their ktlint wrapping against files they were never meant to touch. For autocorrecting wrappings this mutates `fileCopy` mid-walk and writes it back via `root.modifiedText = fileCopy.text` (lines 104-106), corrupting whatever later test reuses that cached `KtFile`. It also feeds Defect 1 (inflating `active.size`) and quietly re-walks every file with every ever-tested rule — ironic for a perf change.

The guard comment at lines 84-90 only defends against *provider-registered* stale rules (`isActiveByConfig`); it does nothing about *previously-triggered* ones, which are the ones that bite in tests.

## Why it's intermittent

Within a single test JVM (Gradle uses `maxParallelForks` = separate JVMs, but classes within a fork run sequentially), the outcome depends entirely on **which tests ran before** and populated the global `registry`/`contexts`. JUnit's default method/class ordering isn't a stable contract, so the same suite passes or fails depending on ordering — the classic shared-singleton flake.

## Recommended fix

The clean fix is to stop relying on JVM-global accumulation:

1. **Make eviction robust** — don't tie it to `active.size`. Either evict purely by identity when a rule is done (and let `computeIfAbsent` rebuild on demand), or track the *expected* visitor count for the file rather than the registry size. Simplest correct option: drop the counter and evict on the rule that observes its own findings, or rebuild whenever `modifiedText` changed since the cached build.
2. **Scope the registry per run, not per JVM** — clear it between analyses, and in tests add an `@AfterEach`/`@BeforeEach` that resets `KtlintEngine` (a small `internal fun reset()` clearing `registry` and `contexts`). That removes both the stale-rule participation and the inflated counter.

Both are needed: (1) stops stale contexts being served, (2) stops stale rules participating.

Want me to implement this on `claude/gracious-brahmagupta-bfcw8u`? I'd add an `internal fun reset()` to `KtlintEngine`, wire it into the ktlint-wrapper test lifecycle, and fix the eviction-counter logic so the per-file context is rebuilt when its inputs change — then run `./gradlew :detekt-rules-ktlint-wrapper:test` to confirm the flake is gone.
</details>

## Your Environment

* Version of detekt used: main
* Version of Gradle used (if applicable): 9.5.1
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version:
* Link to your project (if it's a public repository):
